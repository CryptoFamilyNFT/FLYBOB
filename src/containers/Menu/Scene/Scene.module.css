.Scene {
  border: 6px solid aqua;
  border-radius: 4px;
  height: 60vh;
}

.Modal {
  height: 100%;
  width: 100%;
  background-color: rgba(0,0,0, 0.8);
  display: flex;
  justify-content: center;
  color: white
}

.modal {
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: rgba(0,0,0, 0.8);
  border: none;
  height: 100%;
    width: 100%;
}

.paper {
  background-color: rgba(0,0,0, 0.0);
  padding: 20px; /* Cambiato da theme.spacing(2, 4, 3) */
  text-align: center;
  border: 2px solid transparent;
}

.gameOverImage {
  max-width: 100%;
  height: auto;
}

.gameOverButton {
  margin-top: 20px; /* Cambiato da theme.spacing(2) */
}


@media screen and (min-height: 667px) {
  .Scene {
    height: 64vh;
  }
}

@media (min-height: 731px) {
  .Scene {
    height: fit-content;
  }

  .Modal {
    height: 100%;
    width: 100%;
  }

  .GameOver {
    height: 100%;
    width: 100%;
    justify-content: center;
  }
}
