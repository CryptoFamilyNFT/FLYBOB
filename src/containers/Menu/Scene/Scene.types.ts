export type CoordinatesType = {
  x: number;
  y: number;
  rotation?: number;
};
