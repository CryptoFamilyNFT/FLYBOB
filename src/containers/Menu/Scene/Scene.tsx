import React, {useEffect, useRef} from 'react';
import classes from './Scene.module.css';
import _cloneDeep from 'lodash/cloneDeep';

import {
  sounds,
  sceneImages,
  GAP_VALUE,
  GRAVITATION_VALUE,
  BIRD,
  PIPE_X_STOP_POINT,
  CANVAS
} from './constants';
import {CoordinatesType} from './Scene.types';

const BIRD_COPY = _cloneDeep(BIRD);

let backgroundImage: HTMLImageElement;
let birdImage: HTMLImageElement;
let groundImage: HTMLImageElement;
let pipeNorthImage: HTMLImageElement;
let pipeSouthImage: HTMLImageElement;
let flySound: HTMLAudioElement;
let scoreSound: HTMLAudioElement

const Scene = () => {
    const canvasRef = useRef<HTMLCanvasElement>(null);
    const canvasContext = useRef<CanvasRenderingContext2D>();
    let pipesCoordinate = useRef<CoordinatesType[]>([]);

    let score = 0;
    let requestID: number | undefined;

    const loadSceneImages = async () => {
      backgroundImage = await sceneImages.get('background') as HTMLImageElement;
      birdImage = await sceneImages.get('bird') as HTMLImageElement;
      groundImage = await sceneImages.get('ground') as HTMLImageElement;
      pipeNorthImage = await sceneImages.get('pipeNorth') as HTMLImageElement;
      pipeSouthImage = await sceneImages.get('pipeSouth') as HTMLImageElement;
    };

    const loadSceneSounds = () => {
      flySound = sounds.get('fly') as HTMLAudioElement;
      scoreSound = sounds.get('score') as HTMLAudioElement;
    }

    const initCanvasContext = () => {
      if (canvasRef.current) {
        canvasContext.current = canvasRef.current.getContext('2d') as CanvasRenderingContext2D;
      }
    };

  const setStarterValues = () => {
    BIRD_COPY.position = {
      ...BIRD.position
    };

    score = 0;

    pipesCoordinate.current = [
      {
        x: canvasRef.current?.width as number,
        y: 0,
      }
    ];
  }

  const renderer = () => {
    if (!canvasContext.current || !canvasRef.current) return;
    canvasContext.current.drawImage(
      backgroundImage,
      0,
      0
    );
    for (let i = 0; i < pipesCoordinate.current.length; i++) {
      const constant = pipeNorthImage.height + GAP_VALUE;

      /* Draw top pipe */
      canvasContext.current.drawImage(
        pipeNorthImage,
        pipesCoordinate.current[i].x,
        pipesCoordinate.current[i].y
      );

      /* Draw bottom pipe */
      canvasContext.current.drawImage(
        pipeSouthImage,
        pipesCoordinate.current[i].x,
        pipesCoordinate.current[i].y + constant
      );

      /* Move pipes */
      pipesCoordinate.current[i].x--;

      /* Add new pipes */
      if (pipesCoordinate.current[i].x === PIPE_X_STOP_POINT) {
        pipesCoordinate.current = [
          ...pipesCoordinate.current,
          {
            x: canvasRef.current.width,
            y: Math.floor(Math.random() * pipeNorthImage.height) - pipeNorthImage.height,
          }
        ];
      }

      /* Bit about pipes */
      if (
        (
          BIRD_COPY.position.x + birdImage.width >= pipesCoordinate.current[i].x &&
          BIRD_COPY.position.x <= pipesCoordinate.current[i].x + pipeNorthImage.width &&
          (
            BIRD_COPY.position.y <= pipesCoordinate.current[i].y + pipeNorthImage.height ||
            BIRD_COPY.position.y + birdImage.height >= pipesCoordinate.current[i].y + constant
          )
        ) ||
        BIRD_COPY.position.y + birdImage.height >= canvasRef.current.height - groundImage.height
      ) {
        if (typeof requestID === "number") {
          window.cancelAnimationFrame(requestID);
        }
        setStarterValues();
        renderer();
      // eslint-disable-next-line no-throw-literal
        throw 'End of the game';
      }

      /* Change score */
      if (pipesCoordinate.current[i].x === 5) {
        score++;
        scoreSound.play();
      }
    }

    /* Draw ground */
    canvasContext.current.drawImage(
      groundImage,
      0,
      canvasRef.current.height - groundImage.height
    );

    /* Draw bird */
    canvasContext.current.drawImage(
      birdImage,
      BIRD_COPY.position.x,
      BIRD_COPY.position.y
    );

    BIRD_COPY.position.y += GRAVITATION_VALUE;

    /* Draw score */
    canvasContext.current.fillStyle = '#000';
    canvasContext.current.font = '16px Century Gothic';
    canvasContext.current.fillText(
      `Score: ${score}`,
      10,
      canvasRef.current.height - 20
    );
    requestID = window.requestAnimationFrame(renderer);
  };

  const removeEventListeners = () => {
    window.removeEventListener('keydown', keyDownPressHandler);
    canvasRef.current?.removeEventListener('touchstart', keyDownPressHandler);
    if (typeof requestID === 'number') {
      window.cancelAnimationFrame(requestID);
    }
  };

  const keyDownPressHandler = () => {
    BIRD_COPY.position.y -= 30;
    flySound.play();
  }

  useEffect(() => {
    initCanvasContext();
    setStarterValues();
    loadSceneSounds();
    (async() => {
      await loadSceneImages();
      renderer();
      window.addEventListener('keydown', keyDownPressHandler);
      canvasRef.current?.addEventListener('touchstart', keyDownPressHandler)
    })();

      return () => {
        removeEventListeners();
      };
    }, [removeEventListeners, renderer, setStarterValues]);

    return (
      <canvas
        className={classes.Scene}
        ref={canvasRef}
        width={CANVAS.width}
        height={CANVAS.height}
      />
    );
};

export default Scene;
