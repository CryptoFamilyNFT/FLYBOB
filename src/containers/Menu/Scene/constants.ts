import {CoordinatesType} from './Scene.types';
import Background from "../../../assets/images/bg.png";
import Bird from "../../../assets/images/bird.png";
import Ground from "../../../assets/images/fg.png";
import TopPipe from "../../../assets/images/pipeNorth.png";
import BottomPipe from "../../../assets/images/pipeSouth.png";
// @ts-ignore
import FlySound from "../../../assets/sounds/fly.mp3";
// @ts-ignore
import ScoreSound from "../../../assets/sounds/score.mp3";
import {createAudio, createImage} from "../../../helpers";

export const GAP_VALUE = 80;
export const GRAVITATION_VALUE = 1.7;

export const CANVAS = {
  width: 288,
  height: 512
}

export const BIRD: { position: CoordinatesType } = {
  position:  {
    x: 10,
    y: 150
  }
};

export const PIPE_X_ENDPOINT = 125;
export const PIPE_X_UPPER_SCORE = 5;

export const sceneImages = new Map([
  [
    'background',
    createImage(Background)
  ],
  [
    'bird',
    createImage(Bird)
  ],
  [
    'ground',
    createImage(Ground)
  ],
  [
    'topPipe',
    createImage(TopPipe)
  ],
  [
    'bottomPipe',
    createImage(BottomPipe)
  ]
]);

export const sounds = new Map([
  [
    'fly',
    createAudio(FlySound)
  ],
  [
    'score', createAudio(ScoreSound)
  ]
]);

export const SCORE_TEXT = {
  fill: '#000',
  font: '16px Century Gothic',
  text: (score: number) => `Score: ${score}`
}
