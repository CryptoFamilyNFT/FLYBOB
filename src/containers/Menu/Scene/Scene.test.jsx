import React from "react";
import renderer from "react-test-renderer";
import Scene from "./Scene";

describe('<Scene />', () => {
  it('should renders correctly', () => {
    const tree = renderer
      .create(<Scene/>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

