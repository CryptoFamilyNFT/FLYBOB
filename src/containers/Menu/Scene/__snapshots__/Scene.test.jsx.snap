// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Scene /> should renders correctly 1`] = `
<canvas
  className="Scene"
  height={512}
  width={288}
/>
`;
